tests/cpp_validation_tests/test_validation.cpp

```diff
@@ -40,238 +86,263 @@ auto read_json(std::filesystem::path const& path) {
-                reinterpret_cast<char const*>(ref_buffer.data) + elements_per_scenario * scenario * component_meta.size;
-            // loop all attribute
-            for (MetaAttribute const& attr : component_meta.attributes) {
+
+    DatasetConst const& result = owning_result.const_dataset.value();
+    auto const& result_info = result.get_info();
+    auto const& result_name = result_info.name();
+    Idx const result_batch_size = result_info.batch_size();
+    auto const& storage = owning_result.storage;
+
+    DatasetConst const& reference_result = owning_reference_result.const_dataset.value();
+    auto const& reference_result_info = reference_result.get_info();
+    auto const& reference_result_name = reference_result_info.name();
+    auto const& reference_storage = owning_reference_result.storage;
+    CHECK(storage.buffers.size() == reference_storage.buffers.size());
+
+    // loop through all scenarios
+    for (Idx scenario_idx{}; scenario_idx < result_batch_size; ++scenario_idx) {
+        // loop through all components
+        for (Idx component_idx{}; component_idx < reference_result_info.n_components(); ++component_idx) {
+            auto const& component_name = reference_result_info.component_name(component_idx);
+            auto const* const component_meta = MetaData::get_component_by_name(reference_result_name, component_name);
+
+            auto const& ref_buffer = reference_storage.buffers.at(component_idx);
+            auto const& buffer = storage.buffers.at(component_idx);
+            Idx const elements_per_scenario = reference_result_info.component_elements_per_scenario(component_idx);
+            CHECK(elements_per_scenario >= 0);
+            // loop through all attributes
+            for (Idx attribute_idx{}; attribute_idx < MetaData::n_attributes(component_meta); ++attribute_idx) {
+                auto const* const attribute_meta = MetaData::get_attribute_by_idx(component_meta, attribute_idx);
+                auto attribute_type = MetaData::attribute_ctype(attribute_meta);
+                auto const& attribute_name = MetaData::attribute_name(attribute_meta);
                 // TODO skip u angle, need a way for common angle
-                if (attr.name == "u_angle"s) {
+                if (attribute_name == "u_angle"s) {
                     continue;
                 }
                 // get absolute tolerance
                 double dynamic_atol = atol.at("default");
                 for (auto const& [reg, value] : atol) {
-                    if (std::regex_match(attr.name, std::regex{reg})) {
+                    if (std::regex_match(attribute_name, std::regex{reg})) {
                         dynamic_atol = value;
                         break;
                     }
                 }
                 // for other _angle attribute, we need to find the magnitue and compare together
                 std::regex const angle_regex("(.*)(_angle)");
                 std::smatch angle_match;
```

Comment:
These two fellas could live out side of the for loops

Reply:
Should they? Isn't it best practice to shorten lifetime of variables as much as possible into their relevant scope? Since both the fellas are used within the for loop, maybe it is best to keep them in? I am not totally sure what the best practice here is though.

Reply:
> Isn't it best practice to shorten lifetime of variables as much as possible into their relevant scope?

yes indeed. however, since the regex is contant, you can also make it `constexpr` (iff std::regex supports that), and/or make it module-level static, in which case it's important that you put it in an anonymous namespace.

Reply:
Not if you reuse them this frequently ;)

Reply:
> since the regex is contant, you can also make it `constexpr` (iff std::regex supports that), and/or make it module-level static

`std::regex` doesn't support `consexpr` and I would avoid making it module-level static because the readability would be heavily affected.

> In this particular case, I would buy in this for the following two reasons:
> 
> * readability
> * relatively cheap construction & destruction

I advocate to leave it as is for these two reasons specially. Taking it to the outter scope really will affect readability, specially in such complex checking logic.

@Jerry-Jinfeng-Guo @mgovers Do we agree to keep it as is in this particular case?

Reply:
Yup, let's leave it like this.